src/prefect/settings.py

```diff
@@ -1793,101 +1412,109 @@ def copy_with_update(
+        included_names = {s.field_name for s in include} if include else None
+        excluded_names = {s.field_name for s in exclude} if exclude else None
 
         if exclude_unset:
-            set_keys = {
-                # Collect all of the "set" keys and cast to `Setting` objects
-                SETTING_VARIABLES[key]
-                for key in self.model_dump(exclude_unset=True)
-            }
-            include.intersection_update(set_keys)
-
-        # Validate the types of items in `include` to prevent exclusion bugs
-        for key in include:
-            if not isinstance(key, Setting):
-                raise TypeError(
-                    "Invalid type {type(key).__name__!r} for key in `include`."
+            if included_names is None:
+                included_names = set(self.model_dump(exclude_unset=True).keys())
+            else:
+                included_names.intersection_update(
+                    {key for key in self.model_dump(exclude_unset=True)}
                 )
 
-        env: dict[str, Any] = self.model_dump(
-            mode="json", include={s.name for s in include}
+        env: Dict[str, Any] = self.model_dump(
+            include=included_names,
+            exclude=excluded_names,
+            mode="json",
+            context={"include_secrets": include_secrets},
         )
+        return {
+            f"{self.model_config.get('env_prefix')}{key.upper()}": str(value)
+            for key, value in env.items()
+            if value is not None
+        }
 
-        # Cast to strings and drop null values
-        return {key: str(value) for key, value in env.items() if value is not None}
+    @model_serializer(
+        mode="wrap", when_used="always"
+    )  # TODO: reconsider `when_used` default for more control
+    def ser_model(
+        self, handler: SerializerFunctionWrapHandler, info: SerializationInfo
+    ) -> Any:
+        ctx = info.context
+        jsonable_self = handler(self)
+        if ctx and ctx.get("include_secrets") is True:
+            dump_kwargs = dict(include=info.include, exclude=info.exclude)
+            jsonable_self.update(
+                {
+                    field_name: visit_collection(
+                        expr=getattr(self, field_name),
+                        visit_fn=partial(handle_secret_render, context=ctx),
+                        return_data=True,
+                    )
+                    for field_name in set(self.model_dump(**dump_kwargs).keys())  # type: ignore
+                }
+            )
+        return jsonable_self
 
-    model_config = ConfigDict(frozen=True)
 
+############################################################################
+# Settings utils
 
 # Functions to instantiate `Settings` instances
 
-_DEFAULTS_CACHE: Optional[Settings] = None
-_FROM_ENV_CACHE: Dict[int, Settings] = {}
+
+def _cast_settings(
+    settings: Union[Dict[Union[str, Setting], Any], Any],
+) -> Dict[Setting, Any]:
```

Comment:
Can you add a doc string to this function? It's not immediately clear what it's used for.

Reply:
yep, added in [c56164d](https://github.com/PrefectHQ/prefect/pull/15412/commits/c56164dc2f0f3aa0680c7138b1f37c8d03376c33)